---
title: "ttBulk - tidyTranscriptomics"
output: github_document
---

[![Build Status](https://travis-ci.org/stemangiola/ttBulk.svg?branch=master)](https://travis-ci.org/stemangiola/ttBulk) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/ttBulk/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/ttBulk?branch=master)

A user-firendly grammar of bulk RNA sequencing data explaration and processing

#  <img src="inst/logo.png" height="139px" width="120px" />

```{r, echo=FALSE, include=FALSE, }
library(knitr)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE, 
                      message = FALSE, cache.lazy = FALSE)

library(tidyverse)
library(magrittr)
library(ttBulk)

my_theme = 	
	theme_bw() +
	theme(
		panel.border = element_blank(),
		axis.line = element_line(),
		panel.grid.major = element_line(size = 0.2),
		panel.grid.minor = element_line(size = 0.1),
		text = element_text(size=12),
		legend.position="bottom",
		aspect.ratio=1,
		strip.background = element_blank(),
		axis.title.x  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
		axis.title.y  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
	)

```

# Introduction

ttBulk is a collection of wrappers for bulk tanscriptomic analyses that follows the "tidy" paradigm. The data structure is a tibble with a column for 

+ sample identifier column
+ transcript identifier column
+ `read count` column
+ annotation (and other info) columns

```{r}
ttBulk::counts # Accessible via ttBulk::counts
```

Every function takes this tructure as input and outputs: (i) this structure with addictional information (action="add"); or (ii) this structure with the isolated new information (action="get")

In brief you can:

+ Going from BAM/SAM to a tidy data frame of counts (FeatureCounts)
+ Adding gene symbols from ensembl IDs
+ Aggregating duplicated gene symbols
+ Adding normalised counts
+ Adding principal components
+ Adding MDS components
+ Rotating principal component or MDS dimensions
+ Running differential transcript abunance analyses (edgeR)
+ Adding batch adjusted read counts (Combat) 
+ Eliminating redunant samples and/or genes
+ Clustering samples and/or genes with kmeans
+ Adding tissue composition (Cibersort)

# Aggregate `transcripts`

Aggregating duplicated transcripts (e.g., isoforms, ensembl). For example, we often have to convert ensembl to gene symbol, in doing so we have to deal with duplicated symbols 

```{r, cache=TRUE}
counts.aggr = 
  ttBulk::counts %>%
  aggregate_duplicates(
  	sample, 
  	transcript, 
  	`read count`,  
  	aggregation_function = sum
  )

counts.aggr

```

# Normalise `read counts`

For visualisation purposes or ad hoc analyses, we may want to calculate the normalised read counts for library size (e.g., with TMM algorithm). These new values will be added to the original data set as `<NAME OF SAMPLE COLUMN> normalised`

```{r, cache=TRUE}
counts.norm =  counts.aggr %>% normalise_counts(sample, transcript, `read count`)
counts.norm
```

We can easily plot the normalised density to check the outcome

```{r}
counts.norm %>% 
	ggplot(aes(`read count normalised` + 1, group=sample, color=`Cell type`)) +
	geom_density() + 
	scale_x_log10() +
	my_theme
```

# Reduce `dimensions` 

For visualisation purposes or ad hoc analyses, we may want to reduce the dimentions of our data, for example using PCA or MDS algorithms. These new values will be added to the original data set.

## MDS

```{r, cache=TRUE}
counts.norm.MDS = 
  counts.norm %>%
  reduce_dimensions(value_column = `read count normalised`, method="MDS" , elements_column = sample, feature_column = transcript, components = 1:10)
counts.norm.MDS
```

```{r, cache=TRUE}
counts.norm.MDS %>% 
	select(contains("Dimension"), sample, `Cell type`) %>%
  distinct() %>%
  GGally::ggpairs(columns = 1:6, ggplot2::aes(colour=`Cell type`)) 


```

## PCA

```{r, cache=TRUE}
counts.norm.PCA = 
  counts.norm %>%
  reduce_dimensions(value_column = `read count normalised`, method="PCA" , elements_column = sample, feature_column = transcript, components = 1:10)
counts.norm.PCA
```

```{r, cache=TRUE}
counts.norm.PCA %>% 
	select(contains("PC"), sample, `Cell type`) %>%
  distinct() %>%
  GGally::ggpairs(columns = 1:6, ggplot2::aes(colour=`Cell type`)) 
```

# Rotate `dimensions`

For visualisation purposes or ad hoc analyses, we may want to rotate the reduced dimentions (or any two numeric columns really) of our data, of a set angle. The rotated dimensions will be added to the original data set as `<NAME OF DIMENSION> rotated <ANGLE>` by default, or as specified in the input arguments.

```{r, cache=TRUE}
counts.norm.MDS.rotated =
  counts.norm.MDS %>%
	rotate_dimensions(`Dimension 1`, `Dimension 2`, rotation_degrees = 45, elements_column = sample)
```

## Original

```{r, cache=TRUE}
counts.norm.MDS.rotated %>%
	distinct(sample, `Dimension 1`,`Dimension 2`, `Cell type`) %>%
	ggplot(aes(x=`Dimension 1`, y=`Dimension 2`, color=`Cell type` )) + 
  geom_point() +
  my_theme
```

## Rotated

```{r, cache=TRUE}
counts.norm.MDS.rotated %>%
	distinct(sample, `Dimension 1 rotated 45`,`Dimension 2 rotated 45`, `Cell type`) %>%
	ggplot(aes(x=`Dimension 1 rotated 45`, y=`Dimension 2 rotated 45`, color=`Cell type` )) + 
  geom_point() +
  my_theme
```

# Differential transcirption

We may want to test for differential transcription between sample-wise factors of interest (e.g., with edgeR). The statistics will be added to the original data. In this example we set action ="get" to just output the non redundant gene statistics.

```{r, cache=TRUE}
ttBulk::counts_mini %>%
	annotate_differential_transcription(
      ~ condition,
      sample_column = sample,
      transcript_column = transcript,
      counts_column = `read count`,
      action="get") 
```

# Adjust `read counts`

Adjust `read counts` for (known) unwanted variation. For visualisation purposes or ad hoc analyses, we may want to adjust our normalised counts to remove known unwanted variation. The adjusted counts will be added to the original data set as `<READ COUNT COLUMN> adjusted`. The formulation is similar to a linear model, where the first covariate is the factor of interest and the second covariate is the unwanted variation. At the moment just an unwanted covariated is allowed at the time.

```{r, cache=TRUE}
counts.norm.adj = 
	counts.norm %>%
	  
	  # Add fake batch and factor of interest
	  left_join( 
	  	(.) %>% 
	  		distinct(sample) %>% 
	  		mutate(batch = sample(0:1, n(), replace = T)) 
	  ) %>%
	 	mutate(factor_of_interest = `Cell type` == "b_cell") %>%
	    
	  # Add covariate
	  adjust_counts(
	  	~ factor_of_interest + batch,	
	  	sample, 
	  	transcript, 
	  	`read count normalised`, 
	  	action = "get"
	  )

counts.norm.adj
```

# Cell type composition

We may want to infer the cell type composition of our samples (e.g., with cibersort). The cell type proportions will be added to the original data. 


```{r, cache=TRUE}
counts.cibersort = 
	ttBulk::counts %>% 
	annotate_cell_type(sample, transcript, `read count`, action="add") 
counts.cibersort
```

We can plot the distributions of cell types across samples, and compare them with the nominal cell type labels to check for the purity of isolation.

```{r, cache=TRUE}
counts.cibersort %>%
  rename(`Cell type experimental` = `Cell type.x`, `Cell type estimated` = `Cell type.y`) %>%
  distinct(sample, `Cell type experimental`, `Cell type estimated`, proportion) %>%
  ggplot(aes(x=`Cell type estimated`, y=proportion, fill=`Cell type experimental`)) + 
  geom_boxplot() + 
  facet_wrap(~`Cell type experimental`) +
  my_theme + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), aspect.ratio=1/5)
```

# Cluster

For visualisation purposes or ad hoc analyses, we may want to cluster our data (e.g., k-means sample-wise). The cluster annotation will be added to the original data set.

## k-means

```{r, cache=TRUE}
counts.norm.cluster = counts.norm %>%
  annotate_clusters(value_column = `read count normalised`, elements_column = sample, feature_column = transcript,	number_of_clusters = 2 )

counts.norm.cluster
```

We can add cluster annotation to the MDS dimesion reduced data set and plot

```{r, cache=TRUE}
 counts.norm.MDS %>%
  annotate_clusters(
  	value_column = `read count normalised`,
  	elements_column = sample, 
  	feature_column = transcript,	
  	number_of_clusters = 2 
  ) %>% 
	distinct(sample, `Dimension 1`, `Dimension 2`, cluster) %>%
	ggplot(aes(x=`Dimension 1`, y=`Dimension 2`, color=cluster)) + 
  geom_point() +
  my_theme
```

# Drop redundant

For visualisation purposes or ad hoc analyses, we may want to remove redundant elements from the original data set (e.g., samples or transcripts). 

## Use correlation or sample removal

```{r, cache=TRUE}
counts.norm.non_redundant = 
	counts.norm.MDS %>% 
  drop_redundant(
  	method = "correlation",
  	elements_column = sample,
  	feature_column = transcript, 
  	value_column = `read count normalised`
  )
```

We can visualise how the reduced redundancy with the reduced dimentions look like

```{r, cache=TRUE}
counts.norm.non_redundant %>%
	distinct(sample, `Dimension 1`, `Dimension 2`, `Cell type`) %>%
	ggplot(aes(x=`Dimension 1`, y=`Dimension 2`, color=`Cell type`)) + 
  geom_point() +
  my_theme

```

## Use reduced dimensions

```{r, cache=TRUE}
counts.norm.non_redundant = 
	counts.norm.MDS %>% 
  drop_redundant(
  	method = "reduced_dimensions",
  	elements_column = sample,
  	feature_column = transcript,
  	Dim_a_column = `Dimension 1`,
  	Dim_b_column = `Dimension 2`
  )
```


We can visualise how the reduced redundancy with the reduced dimentions look like

```{r, cache=TRUE}
counts.norm.non_redundant %>%
	distinct(sample, `Dimension 1`, `Dimension 2`, `Cell type`) %>%
	ggplot(aes(x=`Dimension 1`, y=`Dimension 2`, color=`Cell type`)) + 
  geom_point() +
  my_theme

```

# Other useful wrappers

We can convert a list of BAM/SAM files into a tidy data frame of annotated counts, via FeatureCounts

```{r eval=FALSE}
counts = bam_sam_to_featureCounts_tibble(file_names, genome = "hg38")
```

We can add gene symbols from ensembl identifiers 

```{r eval = F}
counts_ensembl %>% annotate_symbol(ens)
```
