<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparison with base R â€¢ tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparison with base R">
<meta property="og:description" content="tidybulk">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential transcript abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="comparison_with_base_R_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparison with base R</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            
            <h4 class="date">2020-09-26</h4>
      
      
      <div class="hidden name"><code>comparison_with_base_R.Rmd</code></div>

    </div>

    
    
<style type="text/css">
.r {
  max-width: 2800px;!important
  margin-left: auto;
  margin-right: auto;
  padding=0px;!important 
  margin-right=10px;!important 
}
pre{
    padding-left:20px;
}
.column-left{
  float: left;
  width: 50%;
  text-align: left;
}
.column-right{
  float: left;
  width: 50%;
  text-align: left;
}
</style>
<p><!-- badges: start --> <a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <!-- badges: end --></p>
<!---

[![Build Status](https://travis-ci.org/stemangiola/tidybulk.svg?branch=master)](https://travis-ci.org/stemangiola/tidybulk) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/tidybulk/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/tidybulk?branch=master)

-->
<p>In this article we show some examples of the differences in coding between tidybulk/tidyverse and base R. We noted a decrease &gt; 10x of assignments and a decrease of &gt; 2x of line numbers.</p>
<div id="create-tidybulk-tibble-" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidybulk-tibble-" class="anchor"></a>Create <code>tidybulk</code> tibble.</h2>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">tt</span> <span class="op">=</span> <span class="kw">counts_mini</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/tidybulk.html">tidybulk</a></span>(<span class="kw">sample</span>, <span class="kw">transcript</span>, <span class="kw">count</span>)
</pre></div>
</div>
<div id="aggregate-duplicated-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregate-duplicated-transcripts" class="anchor"></a>Aggregate duplicated <code>transcripts</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">tt.aggr</span> <span class="op">=</span> <span class="kw">tt</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/aggregate_duplicates.html">aggregate_duplicates</a></span>()
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">temp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(
    symbol = <span class="kw">dge_list</span><span class="op">$</span><span class="kw">genes</span><span class="op">$</span><span class="kw">symbol</span>,
    <span class="kw">dge_list</span><span class="op">$</span><span class="kw">counts</span>
)
<span class="kw">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/by.html">by</a></span>(<span class="kw">temp</span>, <span class="kw">temp</span><span class="op">$</span><span class="kw">symbol</span>,
    <span class="fu">function</span>(<span class="kw">df</span>)
        <span class="co">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw">df</span>[<span class="fl">1</span>,<span class="fl">1</span>])<span class="op">&gt;</span><span class="fl">0</span>)
            <span class="kw">matrixStats</span>:::<span class="fu">colSums</span>(<span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw">df</span>[,<span class="op">-</span><span class="fl">1</span>]))
)
<span class="kw">dge_list.nr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"rbind"</span>, <span class="kw">dge_list.nr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">dge_list.nr</span>) <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="kw">dge_list</span>)
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="scale-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#scale-counts" class="anchor"></a>Scale <code>counts</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">tt.norm</span> <span class="op">=</span> <span class="kw">tt.aggr</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/identify_abundant.html">identify_abundant</a></span>(factor_of_interest = <span class="kw">condition</span>) <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/scale_abundance.html">scale_abundance</a></span>()
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">library</span>(edgeR)</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a>dgList &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">count_m=</span>x,<span class="dt">group=</span>group)</span>
<span id="cb5-4"><a href="#cb5-4"></a>keep &lt;-<span class="st"> </span><span class="kw">filterByExpr</span>(dgList)</span>
<span id="cb5-5"><a href="#cb5-5"></a>dgList &lt;-<span class="st"> </span>dgList[keep,,keep.lib.sizes=<span class="ot">FALSE</span>]</span>
<span id="cb5-6"><a href="#cb5-6"></a>[...]</span>
<span id="cb5-7"><a href="#cb5-7"></a>dgList &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(dgList, <span class="dt">method=</span><span class="st">"TMM"</span>)</span>
<span id="cb5-8"><a href="#cb5-8"></a>norm_counts.table &lt;-<span class="st"> </span><span class="kw">cpm</span>(dgList)</span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="filter-variable-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-variable-transcripts" class="anchor"></a>Filter <code>variable transcripts</code>
</h2>
<p>We may want to identify and filter variable transcripts.</p>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">tt.norm.variable</span> <span class="op">=</span> <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/keep_variable.html">keep_variable</a></span>()
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span>)

<span class="kw">x</span> <span class="op">=</span> <span class="kw">norm_counts.table</span>

<span class="kw">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>((<span class="kw">x</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="kw">x</span>))<span class="op">^</span><span class="fl">2</span>)
<span class="kw">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="kw">s</span>,decreasing=<span class="fl">TRUE</span>)
<span class="kw">x</span> <span class="op">&lt;-</span> <span class="kw">x</span>[<span class="kw">o</span>[<span class="fl">1L</span><span class="op">:</span><span class="kw">top</span>],,drop=<span class="fl">FALSE</span>]

<span class="kw">norm_counts.table</span> <span class="op">=</span> <span class="kw">norm_counts.table</span>[<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">x</span>)]

<span class="kw">norm_counts.table</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(
        <span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>,
        <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">norm_counts.table</span>)
    ),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce <code>dimensions</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">=</span>
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(method=<span class="st">"MDS"</span>, .dims = <span class="fl">2</span>)
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/limma">limma</a></span>)

<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)
<span class="kw">cmds</span> <span class="op">=</span> <span class="kw">limma</span>::<span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html">plotMDS</a></span>(ndim = <span class="kw">.dims</span>, plot = <span class="fl">FALSE</span>)

<span class="kw">cmds</span> <span class="op">=</span> <span class="kw">cmds</span> <span class="op">%$%</span> 
    <span class="kw">cmdscale.out</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">"Dim%s"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>))

<span class="kw">cmds</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">cmds</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">

</div>
<div id="pca" class="section level3">
<h3 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h3>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="kw">tt.norm.PCA</span> <span class="op">=</span>
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(method=<span class="st">"PCA"</span>, .dims = <span class="fl">2</span>)
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)
<span class="kw">pc</span> <span class="op">=</span> <span class="kw">count_m_log</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(scale = <span class="fl">TRUE</span>)
<span class="kw">variance</span> <span class="op">=</span> <span class="kw">pc</span><span class="op">$</span><span class="kw">sdev</span><span class="op">^</span><span class="fl">2</span>
<span class="kw">variance</span> <span class="op">=</span> (<span class="kw">variance</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="kw">variance</span>))[<span class="fl">1</span><span class="op">:</span><span class="fl">6</span>]
<span class="kw">pc</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">counts</span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">counts</span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">pc</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="tsne" class="section level3">
<h3 class="hasAnchor">
<a href="#tsne" class="anchor"></a>tSNE</h3>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">=</span>
    <span class="kw">breast_tcga_mini</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/tidybulk.html">tidybulk</a></span>(       <span class="kw">sample</span>, <span class="kw">ens</span>, <span class="kw">count_scaled</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/identify_abundant.html">identify_abundant</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        method = <span class="st">"tSNE"</span>,
        perplexity=<span class="fl">10</span>,
        pca_scale =<span class="fl">TRUE</span>
    )
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)

<span class="kw">tsne</span> <span class="op">=</span> <span class="kw">Rtsne</span>::<span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="kw">count_m_log</span>),
    perplexity=<span class="fl">10</span>,
        pca_scale =<span class="fl">TRUE</span>
)<span class="op">$</span><span class="kw">Y</span>
<span class="kw">tsne</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">tsne</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
</div>
<div id="rotate-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#rotate-dimensions" class="anchor"></a>Rotate <code>dimensions</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">=</span>
  <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/rotate_dimensions.html">rotate_dimensions</a></span>(<span class="kw">`Dim1`</span>, <span class="kw">`Dim2`</span>, rotation_degrees = <span class="fl">45</span>, action=<span class="st">"get"</span>)
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">rotation</span> <span class="op">=</span> <span class="fu">function</span>(<span class="kw">m</span>, <span class="kw">d</span>) {
    <span class="kw">r</span> <span class="op">=</span> <span class="kw">d</span> <span class="op">*</span> <span class="kw">pi</span> <span class="op">/</span> <span class="fl">180</span>
    ((<span class="fu"><a href="../reference/arrange.html">bind_rows</a></span>(
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(`1` = <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">cos</a></span>(<span class="kw">r</span>), `2` = <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html">sin</a></span>(<span class="kw">r</span>)),
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(`1` = <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">sin</a></span>(<span class="kw">r</span>), `2` = <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">cos</a></span>(<span class="kw">r</span>))
    ) <span class="op">%&gt;%</span> <span class="kw">as_matrix</span>) <span class="op">%*%</span> <span class="kw">m</span>)
}
<span class="kw">mds_r</span> <span class="op">=</span> <span class="kw">pca</span> <span class="op">%&gt;%</span> <span class="fu">rotation</span>(<span class="kw">rotation_degrees</span>)
<span class="kw">mds_r</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">counts</span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">counts</span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">mds_r</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="test-differential-abundance" class="section level2">
<h2 class="hasAnchor">
<a href="#test-differential-abundance" class="anchor"></a>Test <code>differential abundance</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">tt.de</span> <span class="op">=</span>
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/test_differential_abundance.html">test_differential_abundance</a></span>( <span class="op">~</span> <span class="kw">condition</span>, action=<span class="st">"get"</span>)
<span class="kw">tt.de</span>
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span>)

<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span>(counts=<span class="kw">counts_m</span>,group=<span class="kw">group</span>)
<span class="kw">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span>(<span class="kw">dgList</span>)
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="kw">dgList</span>[<span class="kw">keep</span>,,keep.lib.sizes=<span class="fl">FALSE</span>]
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="kw">dgList</span>)
<span class="kw">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(<span class="op">~</span><span class="kw">group</span>)
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span>(<span class="kw">dgList</span>,<span class="kw">design</span>)
<span class="kw">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span>(<span class="kw">dgList</span>,<span class="kw">design</span>)
<span class="kw">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFTest</a></span>(<span class="kw">fit</span>,coef=<span class="fl">2</span>)
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span>(<span class="kw">qlf</span>, n=<span class="fl">Inf</span>)
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="adjust-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#adjust-counts" class="anchor"></a>Adjust <code>counts</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">tt.norm.adj</span> <span class="op">=</span>
    <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/adjust_abundance.html">adjust_abundance</a></span>(   <span class="op">~</span> <span class="kw">condition</span> <span class="op">+</span> <span class="kw">time</span>)
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">sva</span>)

<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)

<span class="kw">design</span> <span class="op">=</span>
        <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(
            object = <span class="op">~</span> <span class="kw">condition</span> <span class="op">+</span> <span class="kw">time</span>,
            data = <span class="kw">annotation</span>
        )

<span class="kw">count_m_log.sva</span> <span class="op">=</span>
    <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat.html">ComBat</a></span>(
            batch = <span class="kw">design</span>[,<span class="fl">2</span>],
            mod = <span class="kw">design</span>,
            <span class="kw">...</span>
        )

<span class="kw">count_m_log.sva</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(<span class="kw">count_m_log.sva</span>) <span class="op">-</span><span class="fl">1</span>)
<span class="kw">count_m_log.sva</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">counts</span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">counts</span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">count_m_log.sva</span>)),
    <span class="st">"Cell type"</span>
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="deconvolve-cell-type-composition" class="section level2">
<h2 class="hasAnchor">
<a href="#deconvolve-cell-type-composition" class="anchor"></a>Deconvolve <code>Cell type composition</code>
</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">tt.cibersort</span> <span class="op">=</span>
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/deconvolve_cellularity.html">deconvolve_cellularity</a></span>(action=<span class="st">"get"</span>, cores=<span class="fl">1</span>)
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="kw">source</span>(â€˜CIBERSORT.Râ€™)</span>
<span id="cb21-2"><a href="#cb21-2"></a>count_m <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">write.table</span>(<span class="st">"mixture_file.txt"</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a>results &lt;-<span class="st"> </span><span class="kw">CIBERSORT</span>(</span>
<span id="cb21-4"><a href="#cb21-4"></a>    <span class="st">"sig_matrix_file.txt"</span>,</span>
<span id="cb21-5"><a href="#cb21-5"></a>    <span class="st">"mixture_file.txt"</span>,</span>
<span id="cb21-6"><a href="#cb21-6"></a>    <span class="dt">perm=</span><span class="dv">100</span>, <span class="dt">QN=</span><span class="ot">TRUE</span></span>
<span id="cb21-7"><a href="#cb21-7"></a>)</span>
<span id="cb21-8"><a href="#cb21-8"></a>results<span class="op">$</span>cell_type =<span class="st"> </span>tidybulk<span class="op">::</span>counts[</span>
<span id="cb21-9"><a href="#cb21-9"></a>    <span class="kw">match</span>(tidybulk<span class="op">::</span>counts<span class="op">$</span>sample, <span class="kw">rownames</span>(results)),</span>
<span id="cb21-10"><a href="#cb21-10"></a>    <span class="st">"Cell type"</span></span>
<span id="cb21-11"><a href="#cb21-11"></a>]</span></code></pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="cluster-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-samples" class="anchor"></a>Cluster <code>samples</code>
</h2>
<div id="k-means" class="section level3">
<h3 class="hasAnchor">
<a href="#k-means" class="anchor"></a>k-means</h3>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">tt.norm.cluster</span> <span class="op">=</span> <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(method=<span class="st">"kmeans"</span>, centers = <span class="fl">2</span>, action=<span class="st">"get"</span> )
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">count_m_log</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="kw">count_m</span> <span class="op">+</span> <span class="fl">1</span>)

<span class="kw">k</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span>(<span class="kw">count_m_log</span>, iter.max = <span class="fl">1000</span>, <span class="kw">...</span>)
<span class="kw">cluster</span> <span class="op">=</span> <span class="kw">k</span><span class="op">$</span><span class="kw">cluster</span>

<span class="kw">cluster</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">cluster</span>)),
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Cell type"</span>, <span class="st">"Dim1"</span>, <span class="st">"Dim2"</span>)
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="snn" class="section level3">
<h3 class="hasAnchor">
<a href="#snn" class="anchor"></a>SNN</h3>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">=</span>
    <span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(method = <span class="st">"SNN"</span>)
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://satijalab.org/seurat">Seurat</a></span>)

<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="kw">count_m</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(
    <span class="kw">snn</span>, display.progress = <span class="fl">TRUE</span>,
    num.cores=<span class="fl">4</span>, do.par = <span class="fl">TRUE</span>
)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">snn</span>, selection.method = <span class="st">"vst"</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">snn</span>, selection.method = <span class="st">"vst"</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="kw">snn</span>, npcs = <span class="fl">30</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="kw">snn</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="kw">snn</span>, method = <span class="st">"igraph"</span>, <span class="kw">...</span>)
<span class="kw">snn</span> <span class="op">=</span> <span class="kw">snn</span>[[<span class="st">"seurat_clusters"</span>]]

<span class="kw">snn</span><span class="op">$</span><span class="kw">cell_type</span> <span class="op">=</span> <span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span>[
    <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw">tidybulk</span>::<span class="kw"><a href="../reference/counts.html">counts</a></span><span class="op">$</span><span class="kw">sample</span>, <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span>(<span class="kw">snn</span>)),
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Cell type"</span>, <span class="st">"Dim1"</span>, <span class="st">"Dim2"</span>)
]
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
</div>
<div id="drop-redundant-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#drop-redundant-transcripts" class="anchor"></a>Drop <code>redundant</code> transcripts</h2>
<div class="column-left">
<p>Tidy transcriptomics</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">=</span>
    <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/remove_redundancy.html">remove_redundancy</a></span>(    method = <span class="st">"correlation"</span> )
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://github.com/dgrtwo/widyr">widyr</a></span>)

<span class="kw">.data.correlated</span> <span class="op">=</span>
    <span class="fu"><a href="https://rdrr.io/pkg/widyr/man/pairwise_cor.html">pairwise_cor</a></span>(
        <span class="kw">counts</span>,
        <span class="kw">sample</span>,
        <span class="kw">transcript</span>,
        <span class="kw">rc</span>,
        sort = <span class="fl">TRUE</span>,
        diag = <span class="fl">FALSE</span>,
        upper = <span class="fl">FALSE</span>
    ) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/filter.html">filter</a></span>(<span class="kw">correlation</span> <span class="op">&gt;</span> <span class="kw">correlation_threshold</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/distinct.html">distinct</a></span>(<span class="kw">item1</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/rename.html">rename</a></span>(<span class="op">!</span><span class="op">!</span><span class="kw">.element</span> <span class="op">:=</span> <span class="kw">item1</span>)

<span class="co"># Return non redundant data frame</span>
<span class="kw">counts</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html">anti_join</a></span>(<span class="kw">.data.correlated</span>) <span class="op">%&gt;%</span>
    <span class="fu">spread</span>(<span class="kw">sample</span>, <span class="kw">rc</span>, <span class="op">-</span> <span class="kw">transcript</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/left_join.html">left_join</a></span>(<span class="kw">annotation</span>)
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="draw-heatmap" class="section level2">
<h2 class="hasAnchor">
<a href="#draw-heatmap" class="anchor"></a>Draw <code>heatmap</code>
</h2>
<div class="column-left">
<p>tidytranscriptomics</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>

  <span class="co"># filter lowly abundant</span>
  <span class="fu"><a href="../reference/keep_abundant.html">keep_abundant</a></span>() <span class="op">%&gt;%</span>

  <span class="co"># extract 500 most variable genes</span>
  <span class="fu"><a href="../reference/keep_variable.html">keep_variable</a></span>( .abundance = <span class="kw">count_scaled</span>, top = <span class="fl">500</span>) <span class="op">%&gt;%</span>

  <span class="co"># create heatmap</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span>(<span class="kw">sample</span>, <span class="kw">transcript</span>, <span class="kw">count_scaled</span>, transform = <span class="kw">log1p</span>) <span class="op">%&gt;%</span>
    <span class="fu">add_tile</span>(<span class="kw">`Cell type`</span>) 
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="co"># Example taken from airway dataset from BioC2020 workshop. </span>
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span>(<span class="kw">airway</span>)
<span class="kw">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw">dgList</span><span class="op">$</span><span class="kw">samples</span><span class="op">$</span><span class="kw">`Cell type`</span>)
<span class="kw">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span>(<span class="kw">dgList</span>, group=<span class="kw">group</span>)
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="kw">dgList</span>[<span class="kw">keep.exprs</span>,, keep.lib.sizes=<span class="fl">FALSE</span>]
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="kw">dgList</span>)
<span class="kw">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span>(<span class="kw">dgList</span>, log=<span class="fl">TRUE</span>)
<span class="kw">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="kw">logcounts</span>, <span class="fl">1</span>, <span class="kw">var</span>)
<span class="kw">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw">var_genes</span>, decreasing=<span class="fl">TRUE</span>))[<span class="fl">1</span><span class="op">:</span><span class="fl">500</span>]
<span class="kw">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="kw">logcounts</span>[<span class="kw">select_var</span>,]
<span class="kw">colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#fefada"</span> )
<span class="kw">col.group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>,<span class="st">"grey"</span>)[<span class="kw">group</span>]
<span class="kw">gplots</span>::<span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span>(<span class="kw">highly_variable_lcpm</span>, col=<span class="kw">colours</span>, trace=<span class="st">"none"</span>, ColSideColors=<span class="kw">col.group</span>, scale=<span class="st">"row"</span>)
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="draw-density-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#draw-density-plot" class="anchor"></a>Draw <code>density plot</code>
</h2>
<div class="column-left">
<p>tidytranscriptomics</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="co"># Example taken from airway dataset from BioC2020 workshop. </span>
<span class="kw">airway</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/tidybulk.html">tidybulk</a></span>() <span class="op">%&gt;%</span>
      <span class="fu"><a href="../reference/identify_abundant.html">identify_abundant</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/scale_abundance.html">scale_abundance</a></span>() <span class="op">%&gt;%</span>
    <span class="fu">pivot_longer</span>(cols = <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span>(<span class="st">"counts"</span>), names_to = <span class="st">"source"</span>, values_to = <span class="st">"abundance"</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/filter.html">filter</a></span>(<span class="op">!</span><span class="kw">lowly_abundant</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">abundance</span> <span class="op">+</span> <span class="fl">1</span>, color=<span class="kw">sample</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span><span class="kw">source</span>) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() 
</pre></div>
</div>
<div class="column-right">
<p>Base R</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="co"># Example taken from airway dataset from BioC2020 workshop. </span>
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html">SE2DGEList</a></span>(<span class="kw">airway</span>)
<span class="kw">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw">dgList</span><span class="op">$</span><span class="kw">samples</span><span class="op">$</span><span class="kw">dex</span>)
<span class="kw">keep.exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span>(<span class="kw">dgList</span>, group=<span class="kw">group</span>)
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="kw">dgList</span>[<span class="kw">keep.exprs</span>,, keep.lib.sizes=<span class="fl">FALSE</span>]
<span class="kw">dgList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span>(<span class="kw">dgList</span>)
<span class="kw">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span>(<span class="kw">dgList</span>, log=<span class="fl">TRUE</span>)
<span class="kw">var_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="kw">logcounts</span>, <span class="fl">1</span>, <span class="kw">var</span>)
<span class="kw">select_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw">var_genes</span>, decreasing=<span class="fl">TRUE</span>))[<span class="fl">1</span><span class="op">:</span><span class="fl">500</span>]
<span class="kw">highly_variable_lcpm</span> <span class="op">&lt;-</span> <span class="kw">logcounts</span>[<span class="kw">select_var</span>,]
<span class="kw">colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#440154FF"</span>, <span class="st">"#21908CFF"</span>, <span class="st">"#fefada"</span> )
<span class="kw">col.group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>,<span class="st">"grey"</span>)[<span class="kw">group</span>]
<span class="kw">gplots</span>::<span class="fu"><a href="https://rdrr.io/pkg/gplots/man/heatmap.2.html">heatmap.2</a></span>(<span class="kw">highly_variable_lcpm</span>, col=<span class="kw">colours</span>, trace=<span class="st">"none"</span>, ColSideColors=<span class="kw">col.group</span>, scale=<span class="st">"row"</span>)
</pre></div>
</div>
<div style="clear:both;">

</div>
</div>
<div id="appendix" class="section level2">
<h2 class="hasAnchor">
<a href="#appendix" class="anchor"></a>Appendix</h2>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
</pre></div>
<pre><code>## R version 4.0.2 Patched (2020-09-24 r79253)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] tidybulk_1.1.8   ggrepel_0.8.2    ggplot2_3.3.2    magrittr_1.5    
## [5] tibble_3.0.3     tidyr_1.1.2      dplyr_1.0.2      knitr_1.30      
## [9] BiocStyle_2.17.1
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.1.10            tidytext_0.2.6             
##   [3] systemfonts_0.3.1           plyr_1.8.6                 
##   [5] igraph_1.2.5                lazyeval_0.2.2             
##   [7] splines_4.0.2               BiocParallel_1.23.2        
##   [9] listenv_0.8.0               SnowballC_0.7.0            
##  [11] GenomeInfoDb_1.25.11        sva_3.37.0                 
##  [13] digest_0.6.25               htmltools_0.5.0            
##  [15] fansi_0.4.1                 memoise_1.1.0              
##  [17] tensor_1.5                  cluster_2.1.0              
##  [19] ROCR_1.0-11                 limma_3.45.14              
##  [21] globals_0.13.0              readr_1.3.1                
##  [23] annotate_1.67.1             matrixStats_0.56.0         
##  [25] pkgdown_1.6.1.9000          colorspace_1.4-1           
##  [27] blob_1.2.1                  rappdirs_0.3.1             
##  [29] xfun_0.17                   crayon_1.3.4               
##  [31] RCurl_1.98-1.2              jsonlite_1.7.1             
##  [33] spatstat.data_1.4-3         spatstat_1.64-1            
##  [35] genefilter_1.71.0           survival_3.2-3             
##  [37] zoo_1.8-8                   glue_1.4.2                 
##  [39] polyclip_1.10-0             gtable_0.3.0               
##  [41] zlibbioc_1.35.0             XVector_0.29.3             
##  [43] leiden_0.3.3                DelayedArray_0.15.8        
##  [45] future.apply_1.6.0          BiocGenerics_0.35.4        
##  [47] abind_1.4-5                 scales_1.1.1               
##  [49] DBI_1.1.0                   cpp11_0.2.1                
##  [51] edgeR_3.31.4                miniUI_0.1.1.1             
##  [53] Rcpp_1.0.5                  widyr_0.1.3                
##  [55] viridisLite_0.3.0           xtable_1.8-4               
##  [57] reticulate_1.16             bit_4.0.4                  
##  [59] rsvd_1.0.3                  preprocessCore_1.51.0      
##  [61] stats4_4.0.2                htmlwidgets_1.5.1          
##  [63] httr_1.4.2                  RColorBrewer_1.1-2         
##  [65] ellipsis_0.3.1              Seurat_3.2.1               
##  [67] ica_1.0-2                   pkgconfig_2.0.3            
##  [69] XML_3.99-0.5                uwot_0.1.8                 
##  [71] deldir_0.1-29               locfit_1.5-9.4             
##  [73] utf8_1.1.4                  tidyselect_1.1.0           
##  [75] rlang_0.4.7                 reshape2_1.4.4             
##  [77] later_1.1.0.1               AnnotationDbi_1.51.3       
##  [79] munsell_0.5.0               tools_4.0.2                
##  [81] cli_2.0.2                   generics_0.0.2             
##  [83] RSQLite_2.2.0               broom_0.7.0                
##  [85] ggridges_0.5.2              evaluate_0.14              
##  [87] stringr_1.4.0               fastmap_1.0.1              
##  [89] goftest_1.2-2               yaml_2.2.1                 
##  [91] ragg_0.3.1                  bit64_4.0.5                
##  [93] fs_1.5.0                    fitdistrplus_1.1-1         
##  [95] purrr_0.3.4                 RANN_2.6.1                 
##  [97] pbapply_1.4-3               future_1.19.1              
##  [99] nlme_3.1-149                mime_0.9                   
## [101] tokenizers_0.2.1            compiler_4.0.2             
## [103] plotly_4.9.2.1              png_0.1-7                  
## [105] e1071_1.7-3                 spatstat.utils_1.17-0      
## [107] stringi_1.5.3               desc_1.2.0                 
## [109] lattice_0.20-41             Matrix_1.2-18              
## [111] vctrs_0.3.4                 pillar_1.4.6               
## [113] lifecycle_0.2.0             BiocManager_1.30.10        
## [115] lmtest_0.9-38               RcppAnnoy_0.0.16           
## [117] data.table_1.13.0           cowplot_1.1.0              
## [119] bitops_1.0-6                irlba_2.3.3                
## [121] httpuv_1.5.4                patchwork_1.0.1            
## [123] GenomicRanges_1.41.6        R6_2.4.1                   
## [125] bookdown_0.20               promises_1.1.1             
## [127] gridExtra_2.3               KernSmooth_2.23-17         
## [129] janeaustenr_0.1.5           IRanges_2.23.10            
## [131] codetools_0.2-16            MASS_7.3-53                
## [133] assertthat_0.2.1            SummarizedExperiment_1.19.6
## [135] rprojroot_1.3-2             withr_2.3.0                
## [137] sctransform_0.3             S4Vectors_0.27.13          
## [139] GenomeInfoDbData_1.2.3      mgcv_1.8-33                
## [141] parallel_4.0.2              hms_0.5.3                  
## [143] rpart_4.1-15                grid_4.0.2                 
## [145] class_7.3-17                rmarkdown_2.3              
## [147] Rtsne_0.15                  Biobase_2.49.1             
## [149] shiny_1.5.0</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
