<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the tidybulk package • tidybulk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the tidybulk package">
<meta property="og:description" content="tidybulk">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="introduction_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the tidybulk package</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            
            <h4 class="date">2020-08-08</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- <style type="text/css"> -->
<!-- .main-container { -->
<!--   max-width: 2800px;!important -->
<!--   margin-left: auto; -->
<!--   margin-right: auto; -->
<!-- } -->
<!-- </style> -->
<style>
.column-left{
  float: left;
  width: 50%;
  text-align: left;
}
.column-right{
  float: right;
  width: 50%;
  text-align: right;
}
</style>
<p><strong>Brings transcriptomics to the tidyverse!</strong></p>
<div id="section" class="section level1">
<h1 class="hasAnchor">
<a href="#section" class="anchor"></a><img src="../inst/logo.png" height="139px" width="120px">
</h1>
<p><!-- badges: start --> <a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <!-- badges: end --></p>
<!---

[![Build Status](https://travis-ci.org/stemangiola/tidybulk.svg?branch=master)](https://travis-ci.org/stemangiola/tidybulk) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/tidybulk/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/tidybulk?branch=master)

-->
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>aggregate_duplicates</code></td>
<td>Aggregate robustly abundance and annotation of duplicated transcripts</td>
</tr>
<tr class="even">
<td><code>scale_abundance</code></td>
<td>Scale (normalise) abundance for RNA requencing depth</td>
</tr>
<tr class="odd">
<td><code>reduce_dimensions</code></td>
<td>Perform dimensionality reduction (PCA, MDS, tSNE)</td>
</tr>
<tr class="even">
<td><code>cluster_elements</code></td>
<td>Labels elements with cluster identity (kmeans, SNN)</td>
</tr>
<tr class="odd">
<td><code>remove_redundancy</code></td>
<td>Filter out elements with highly correlated features</td>
</tr>
<tr class="even">
<td><code>adjust_abundance</code></td>
<td>Remove known unwanted variation (Combat)</td>
</tr>
<tr class="odd">
<td><code>test_differential_abundance</code></td>
<td>Differenatial abundance testing (DE)</td>
</tr>
<tr class="even">
<td><code>deconvolve_cellularity</code></td>
<td>Estimated tissue composition (Cibersort)</td>
</tr>
<tr class="odd">
<td><code>keep_variable</code></td>
<td>Filter top variable features</td>
</tr>
<tr class="even">
<td><code>keep_abundant</code></td>
<td>Filter out lowly abundant transcripts</td>
</tr>
<tr class="odd">
<td><code>test_gene_enrichment</code></td>
<td>Gene enrichment analyses (EGSEA)</td>
</tr>
<tr class="even">
<td><code>test_gene_overrepresentation</code></td>
<td>Gene enrichment on list of transcript names (no rank)</td>
</tr>
<tr class="odd">
<td><code>impute_abundance</code></td>
<td>Impute abundance for missing data points using sample groupings</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidybulk</code></td>
<td>add tidybulk attributes to a tibble object</td>
</tr>
<tr class="even">
<td><code>tidybulk_SAM_BAM</code></td>
<td>Convert SAM BAM files into tidybulk tibble</td>
</tr>
<tr class="odd">
<td><code>pivot_sample</code></td>
<td>Select sample-wise columns/information</td>
</tr>
<tr class="even">
<td><code>pivot_transcript</code></td>
<td>Select transcript-wise columns/information</td>
</tr>
<tr class="odd">
<td><code>rotate_dimensions</code></td>
<td>Rotate two dimensions of a degree</td>
</tr>
<tr class="even">
<td><code>ensembl_to_symbol</code></td>
<td>Add gene symbol from ensembl IDs</td>
</tr>
<tr class="odd">
<td><code>symbol_to_entrez</code></td>
<td>Add entrez ID from gene symbol</td>
</tr>
</tbody>
</table>
</div>
<div id="minimal-input-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#minimal-input-data-frame" class="anchor"></a>Minimal input data frame</h2>
<table class="table">
<thead><tr class="header">
<th>sample</th>
<th>transcript</th>
<th>abundance</th>
<th>annotation</th>
</tr></thead>
<tbody><tr class="odd">
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td><code>integer</code></td>
<td>…</td>
</tr></tbody>
</table>
</div>
<div id="output-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#output-data-frame" class="anchor"></a>Output data frame</h2>
<table class="table">
<thead><tr class="header">
<th>sample</th>
<th>transcript</th>
<th>abundance</th>
<th>annotation</th>
<th>new information</th>
</tr></thead>
<tbody><tr class="odd">
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td>
<code>chr</code> or <code>fctr</code>
</td>
<td><code>integer</code></td>
<td>…</td>
<td>…</td>
</tr></tbody>
</table>
</div>
<div id="create-tidybulk-tibble--it-memorises-key-column-names" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidybulk-tibble--it-memorises-key-column-names" class="anchor"></a>Create <code>tidybulk</code> tibble. It memorises key column names</h2>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">tt</span> <span class="op">=</span> <span class="kw">counts_mini</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/tidybulk.html">tidybulk</a></span>(<span class="kw">sample</span>, <span class="kw">transcript</span>, <span class="kw">count</span>)
</pre></div>
<p><strong>All tidybulk methods are directly compatible with SummarizedExperiment as well.</strong></p>
</div>
<div id="aggregate-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregate-transcripts" class="anchor"></a>Aggregate <code>transcripts</code>
</h2>
<p>tidybulk provide the <code>aggregate_duplicates</code> function to aggregate duplicated transcripts (e.g., isoforms, ensembl). For example, we often have to convert ensembl symbols to gene/transcript symbol, but in doing so we have to deal with duplicates. <code>aggregate_duplicates</code> takes a tibble and column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) as arguments and returns a tibble with aggregate transcript with the same name. All the rest of the column are appended, and factors and boolean are appended as characters.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">tt.aggr</span> <span class="op">=</span>  <span class="kw">tt</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/aggregate_duplicates.html">aggregate_duplicates</a></span>(     aggregation_function = <span class="kw">sum</span> )

<span class="kw">tt.aggr</span>
</pre></div>
<pre><code>## # A tibble: 2,635 x 7
##    sample     transcript `Cell type` count time  condition `merged transcripts`
##    &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;                    &lt;dbl&gt;
##  1 SRR1740034 TNFRSF4    b_cell          6 0 d   TRUE                         1
##  2 SRR1740034 PLCH2      b_cell        926 0 d   TRUE                         1
##  3 SRR1740034 PADI4      b_cell         21 0 d   TRUE                         1
##  4 SRR1740034 PAX7       b_cell          0 0 d   TRUE                         1
##  5 SRR1740034 CDA        b_cell          1 0 d   TRUE                         1
##  6 SRR1740034 RCAN3      b_cell        905 0 d   TRUE                         1
##  7 SRR1740034 SMPDL3B    b_cell          3 0 d   TRUE                         1
##  8 SRR1740034 EPB41      b_cell       4667 0 d   TRUE                         1
##  9 SRR1740034 LCK        b_cell        436 0 d   TRUE                         1
## 10 SRR1740034 COL8A2     b_cell          1 0 d   TRUE                         1
## # … with 2,625 more rows</code></pre>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">se.aggr</span> <span class="op">=</span>  <span class="kw">se_mini</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/aggregate_duplicates.html">aggregate_duplicates</a></span>(    aggregation_function = <span class="kw">sum</span> )

<span class="kw">se.aggr</span>
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(1): count
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(0):
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(4): Cell.type time condition merged.transcripts</code></pre>
</div>
<div id="scale-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#scale-counts" class="anchor"></a>Scale <code>counts</code>
</h2>
<p>We may want to compensate for sequencing depth, scaling the transcript abundance (e.g., with TMM algorithm, Robinson and Oshlack doi.org/10.1186/gb-2010-11-3-r25). <code>scale_abundance</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a method as arguments and returns a tibble with additional columns with scaled data as <code>&lt;NAME OF COUNT COLUMN&gt;_scaled</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">tt.norm</span> <span class="op">=</span>  <span class="kw">tt.aggr</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/scale_abundance.html">scale_abundance</a></span>(method=<span class="st">"TMM"</span>)

<span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw">`count`</span>, <span class="kw">count_scaled</span>, <span class="kw">lowly_abundant</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span>())
</pre></div>
<pre><code>## # A tibble: 2,635 x 11
##    count count_scaled lowly_abundant sample transcript `Cell type` time 
##    &lt;dbl&gt;        &lt;dbl&gt; &lt;lgl&gt;          &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt;
##  1  1035      1900.   TRUE           SRR17… ABCB4      b_cell      0 d  
##  2    45        82.6  TRUE           SRR17… ABCB9      b_cell      0 d  
##  3  7151     13129.   FALSE          SRR17… ACAP1      b_cell      0 d  
##  4     2         3.67 TRUE           SRR17… ACHE       b_cell      0 d  
##  5  2278      4182.   FALSE          SRR17… ACP5       b_cell      0 d  
##  6 11156     20482.   TRUE           SRR17… ADAM28     b_cell      0 d  
##  7    72       132.   TRUE           SRR17… ADAMDEC1   b_cell      0 d  
##  8     0         0    TRUE           SRR17… ADAMTS3    b_cell      0 d  
##  9   298       547.   FALSE          SRR17… ADRB2      b_cell      0 d  
## 10     8        14.7  TRUE           SRR17… AIF1       b_cell      0 d  
## # … with 2,625 more rows, and 4 more variables: condition &lt;chr&gt;, `merged
## #   transcripts` &lt;dbl&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;</code></pre>
<p>We can easily plot the scaled density to check the scaling outcome. On the x axis we have the log scaled counts, on the y axes we have the density, data is grouped by sample and coloured by cell type.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">count_scaled</span> <span class="op">+</span> <span class="fl">1</span>, group=<span class="kw">sample</span>, color=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() <span class="op">+</span>
    <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/plot_normalise-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">se.norm</span> <span class="op">=</span>  <span class="kw">se.aggr</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/scale_abundance.html">scale_abundance</a></span>(method=<span class="st">"TMM"</span>)

<span class="kw">se.norm</span>
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(6): Cell.type time ... TMM multiplier</code></pre>
</div>
<div id="filter-variable-transcripts" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-variable-transcripts" class="anchor"></a>Filter <code>variable transcripts</code>
</h2>
<p>We may want to identify and filter variable transcripts.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">tt.norm.variable</span> <span class="op">=</span> <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/keep_variable.html">keep_variable</a></span>()
</pre></div>
</div>
<div id="reduce-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce <code>dimensions</code>
</h2>
<p>We may want to reduce the dimensions of our data, for example using PCA or MDS algorithms. <code>reduce_dimensions</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a method (e.g., MDS or PCA) as arguments and returns a tibble with additional columns for the reduced dimensions.</p>
<p><strong>MDS</strong> (Robinson et al., 10.1093/bioinformatics/btp616)</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">=</span>  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(.abundance = <span class="kw">count_scaled</span>, method=<span class="st">"MDS"</span>, .dims = <span class="fl">3</span>)

<span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"Dim"</span>), <span class="kw">`Cell type`</span>, <span class="kw">time</span> ) <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/distinct.html">distinct</a></span>()
</pre></div>
<pre><code>## # A tibble: 5 x 6
##   sample       Dim1    Dim2       Dim3 `Cell type`       time 
##   &lt;chr&gt;       &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;             &lt;chr&gt;
## 1 SRR1740034 -0.882 -0.708  -0.0291    b_cell            0 d  
## 2 SRR1740035 -0.870 -0.715   0.0000434 b_cell            1 d  
## 3 SRR1740043  1.40   0.0479 -0.471     monocyte          1 d  
## 4 SRR1740058 -0.935  1.39    0.00575   t_cell            0 d  
## 5 SRR1740067  1.29  -0.0121  0.494     dendritic_myeloid 1 d</code></pre>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data is coloured by cell type.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"Dim"</span>), <span class="kw">sample</span>, <span class="kw">`Cell type`</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/distinct.html">distinct</a></span>() <span class="op">%&gt;%</span>
  <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">`Cell type`</span>))
</pre></div>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">se.norm.MDS</span> <span class="op">=</span>  <span class="kw">se.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(.abundance = <span class="kw">count_scaled</span>, method=<span class="st">"MDS"</span>, .dims = <span class="fl">3</span>)

<span class="kw">se.norm.MDS</span>
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(9): Cell.type time ... Dim2 Dim3</code></pre>
<p><strong>PCA</strong></p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">tt.norm.PCA</span> <span class="op">=</span> <span class="kw">tt.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(.abundance = <span class="kw">count_scaled</span>, method=<span class="st">"PCA"</span> ,  .dims = <span class="fl">3</span>)

<span class="kw">tt.norm.PCA</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"PC"</span>), <span class="kw">`Cell type`</span>, <span class="kw">time</span> ) <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/distinct.html">distinct</a></span>()
</pre></div>
<pre><code>## # A tibble: 5 x 6
##   sample       PC1    PC2     PC3 `Cell type`       time 
##   &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;             &lt;chr&gt;
## 1 SRR1740034 -7.42 -6.29  -0.487  b_cell            0 d  
## 2 SRR1740035 -7.18 -6.38  -0.0111 b_cell            1 d  
## 3 SRR1740043 11.5   0.643 -5.89   monocyte          1 d  
## 4 SRR1740058 -7.71 12.4    0.244  t_cell            0 d  
## 5 SRR1740067 10.8  -0.386  6.15   dendritic_myeloid 1 d</code></pre>
<p>On the x and y axes axis we have the reduced dimensions 1 to 3, data is coloured by cell type.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="kw">tt.norm.PCA</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"PC"</span>), <span class="kw">sample</span>, <span class="kw">`Cell type`</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/distinct.html">distinct</a></span>() <span class="op">%&gt;%</span>
  <span class="kw">GGally</span>::<span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(columns = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="kw">ggplot2</span>::<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(colour=<span class="kw">`Cell type`</span>))
</pre></div>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">se.norm.PCA</span> <span class="op">=</span> <span class="kw">se.norm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(.abundance = <span class="kw">count_scaled</span>, method=<span class="st">"PCA"</span> ,  .dims = <span class="fl">3</span>)

<span class="kw">se.norm.PCA</span>
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(9): Cell.type time ... PC2 PC3</code></pre>
<p><strong>tSNE</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">=</span>
    <span class="kw">tt_tcga_breast</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method = <span class="st">"tSNE"</span>,
        top = <span class="fl">500</span>,
        perplexity=<span class="fl">10</span>,
        pca_scale =<span class="fl">TRUE</span>
    )
</pre></div>
<pre><code>## Performing PCA
## Read the 251 x 50 data matrix successfully!
## OpenMP is working. 1 threads.
## Using no_dims = 2, perplexity = 10.000000, and theta = 0.500000
## Computing input similarities...
## Building tree...
## Done in 0.01 seconds (sparsity = 0.183394)!
## Learning embedding...
## Iteration 50: error is 66.165586 (50 iterations in 0.04 seconds)
## Iteration 100: error is 67.158930 (50 iterations in 0.04 seconds)
## Iteration 150: error is 68.357932 (50 iterations in 0.04 seconds)
## Iteration 200: error is 69.911419 (50 iterations in 0.04 seconds)
## Iteration 250: error is 67.860294 (50 iterations in 0.04 seconds)
## Iteration 300: error is 1.540732 (50 iterations in 0.04 seconds)
## Iteration 350: error is 1.250115 (50 iterations in 0.03 seconds)
## Iteration 400: error is 1.189901 (50 iterations in 0.03 seconds)
## Iteration 450: error is 1.160104 (50 iterations in 0.03 seconds)
## Iteration 500: error is 1.135350 (50 iterations in 0.03 seconds)
## Iteration 550: error is 1.128585 (50 iterations in 0.03 seconds)
## Iteration 600: error is 1.118486 (50 iterations in 0.03 seconds)
## Iteration 650: error is 1.108773 (50 iterations in 0.03 seconds)
## Iteration 700: error is 1.096734 (50 iterations in 0.03 seconds)
## Iteration 750: error is 1.090017 (50 iterations in 0.03 seconds)
## Iteration 800: error is 1.086164 (50 iterations in 0.03 seconds)
## Iteration 850: error is 1.085002 (50 iterations in 0.03 seconds)
## Iteration 900: error is 1.080266 (50 iterations in 0.03 seconds)
## Iteration 950: error is 1.078875 (50 iterations in 0.03 seconds)
## Iteration 1000: error is 1.078771 (50 iterations in 0.03 seconds)
## Fitting performed in 0.66 seconds.</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"tSNE"</span>, ignore.case = <span class="fl">FALSE</span>), <span class="kw">sample</span>, <span class="kw">Call</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/distinct.html">distinct</a></span>()
</pre></div>
<pre><code>## # A tibble: 251 x 4
##      tSNE1  tSNE2 sample                       Call 
##      &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;                        &lt;fct&gt;
##  1   8.73   -4.55 TCGA-A1-A0SD-01A-11R-A115-07 LumA 
##  2  -6.01    4.90 TCGA-A1-A0SF-01A-11R-A144-07 LumA 
##  3   4.92   -9.92 TCGA-A1-A0SG-01A-11R-A144-07 LumA 
##  4   0.419  -5.60 TCGA-A1-A0SH-01A-11R-A084-07 LumA 
##  5   9.54   -1.31 TCGA-A1-A0SI-01A-11R-A144-07 LumB 
##  6  -6.92   -1.58 TCGA-A1-A0SJ-01A-11R-A084-07 LumA 
##  7 -32.1    10.9  TCGA-A1-A0SK-01A-12R-A084-07 Basal
##  8  -6.02  -12.0  TCGA-A1-A0SM-01A-11R-A084-07 LumA 
##  9  -4.75  -11.3  TCGA-A1-A0SN-01A-11R-A144-07 LumB 
## 10  22.8    -5.18 TCGA-A1-A0SQ-01A-21R-A144-07 LumA 
## # … with 241 more rows</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">`tSNE1`</span>, y = <span class="kw">`tSNE2`</span>, color=<span class="kw">Call</span>)) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span> <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/tsne-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="kw">se.norm.tSNE</span> <span class="op">=</span>
    <span class="kw">se_breast_tcga_mini</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method = <span class="st">"tSNE"</span>,
        top = <span class="fl">500</span>,
        perplexity=<span class="fl">10</span>,
        pca_scale =<span class="fl">TRUE</span>
    )
</pre></div>
<pre><code>## Performing PCA
## Read the 251 x 50 data matrix successfully!
## OpenMP is working. 1 threads.
## Using no_dims = 2, perplexity = 10.000000, and theta = 0.500000
## Computing input similarities...
## Building tree...
## Done in 0.01 seconds (sparsity = 0.183394)!
## Learning embedding...
## Iteration 50: error is 67.959174 (50 iterations in 0.04 seconds)
## Iteration 100: error is 68.236118 (50 iterations in 0.05 seconds)
## Iteration 150: error is 67.266817 (50 iterations in 0.06 seconds)
## Iteration 200: error is 68.298090 (50 iterations in 0.05 seconds)
## Iteration 250: error is 66.373457 (50 iterations in 0.04 seconds)
## Iteration 300: error is 1.577390 (50 iterations in 0.03 seconds)
## Iteration 350: error is 1.197813 (50 iterations in 0.03 seconds)
## Iteration 400: error is 1.152450 (50 iterations in 0.03 seconds)
## Iteration 450: error is 1.125588 (50 iterations in 0.03 seconds)
## Iteration 500: error is 1.117460 (50 iterations in 0.03 seconds)
## Iteration 550: error is 1.114191 (50 iterations in 0.03 seconds)
## Iteration 600: error is 1.110691 (50 iterations in 0.03 seconds)
## Iteration 650: error is 1.108145 (50 iterations in 0.03 seconds)
## Iteration 700: error is 1.100130 (50 iterations in 0.03 seconds)
## Iteration 750: error is 1.095856 (50 iterations in 0.03 seconds)
## Iteration 800: error is 1.093141 (50 iterations in 0.03 seconds)
## Iteration 850: error is 1.093535 (50 iterations in 0.03 seconds)
## Iteration 900: error is 1.093052 (50 iterations in 0.03 seconds)
## Iteration 950: error is 1.092108 (50 iterations in 0.03 seconds)
## Iteration 1000: error is 1.092111 (50 iterations in 0.03 seconds)
## Fitting performed in 0.70 seconds.</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="kw">se.norm.tSNE</span>
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 500 251 
## metadata(0):
## assays(2): count count_scaled
## rownames(500): ENSG00000002834 ENSG00000003989 ... ENSG00000265972
##   ENSG00000272398
## rowData names(0):
## colnames(251): TCGA-A1-A0SD-01A-11R-A115-07
##   TCGA-A1-A0SF-01A-11R-A144-07 ... TCGA-GM-A2DM-01A-11R-A180-07
##   TCGA-GM-A2DN-01A-11R-A180-07
## colData names(3): Call tSNE1 tSNE2</code></pre>
</div>
<div id="rotate-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#rotate-dimensions" class="anchor"></a>Rotate <code>dimensions</code>
</h2>
<p>We may want to rotate the reduced dimensions (or any two numeric columns really) of our data, of a set angle. <code>rotate_dimensions</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and an angle as arguments and returns a tibble with additional columns for the rotated dimensions. The rotated dimensions will be added to the original data set as <code>&lt;NAME OF DIMENSION&gt; rotated &lt;ANGLE&gt;</code> by default, or as specified in the input arguments.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">=</span>
  <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/rotate_dimensions.html">rotate_dimensions</a></span>(<span class="kw">`Dim1`</span>, <span class="kw">`Dim2`</span>, rotation_degrees = <span class="fl">45</span>, .element = <span class="kw">sample</span>)
</pre></div>
<p><strong>Original</strong> On the x and y axes axis we have the first two reduced dimensions, data is coloured by cell type.</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`Cell type`</span> )) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/plot_rotate_1-1.png" width="700"></p>
<p><strong>Rotated</strong> On the x and y axes axis we have the first two reduced dimensions rotated of 45 degrees, data is coloured by cell type.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="kw">tt.norm.MDS.rotated</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">`Dim1 rotated 45`</span>, y=<span class="kw">`Dim2 rotated 45`</span>, color=<span class="kw">`Cell type`</span> )) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/plot_rotate_2-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="kw">se.norm.MDS</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="../reference/rotate_dimensions.html">rotate_dimensions</a></span>(<span class="kw">`Dim1`</span>, <span class="kw">`Dim2`</span>, rotation_degrees = <span class="fl">45</span>, .element = <span class="kw">sample</span>)
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(11): Cell.type time ... Dim1.rotated.45 Dim2.rotated.45</code></pre>
</div>
<div id="test-differential-abundance" class="section level2">
<h2 class="hasAnchor">
<a href="#test-differential-abundance" class="anchor"></a>Test <code>differential abundance</code>
</h2>
<p>We may want to test for differential transcription between sample-wise factors of interest (e.g., with edgeR). <code>test_differential_abundance</code> takes a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a formula representing the desired linear model as arguments and returns a tibble with additional columns for the statistics from the hypothesis test (e.g., log fold change, p-value and false discovery rate).</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="kw">tt</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="../reference/test_differential_abundance.html">test_differential_abundance</a></span>(  <span class="op">~</span> <span class="kw">condition</span>,  action=<span class="st">"only"</span>)
</pre></div>
<pre><code>## # A tibble: 527 x 8
##    transcript  logFC logCPM     F   PValue    FDR significant lowly_abundant
##    &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt;       &lt;lgl&gt;         
##  1 CLEC7A     -11.6   12.7  109.  0.000142 0.0170 TRUE        FALSE         
##  2 HK3        -12.2   13.5   90.1 0.000225 0.0170 TRUE        FALSE         
##  3 APOBEC3A    -8.93  11.1   77.7 0.000319 0.0170 TRUE        FALSE         
##  4 IGSF6       -7.78  10.5   71.8 0.000385 0.0170 TRUE        FALSE         
##  5 RASSF4      -8.38  12.4   69.0 0.000422 0.0170 TRUE        FALSE         
##  6 IL2RA        8.37   9.18  65.5 0.000477 0.0170 TRUE        FALSE         
##  7 TLR8        -9.94  11.6   64.3 0.000497 0.0170 TRUE        FALSE         
##  8 C5AR1       -9.97  12.2   56.7 0.000667 0.0170 TRUE        FALSE         
##  9 FCN1       -12.6   15.4   56.6 0.000670 0.0170 TRUE        FALSE         
## 10 CCR7         8.35  11.9   56.0 0.000687 0.0170 TRUE        FALSE         
## # … with 517 more rows</code></pre>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="kw">se_mini</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/test_differential_abundance.html">test_differential_abundance</a></span>(  <span class="op">~</span> <span class="kw">condition</span>)
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(1): count
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(7): logFC logCPM ... significant lowly_abundant
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(3): Cell.type time condition</code></pre>
</div>
<div id="adjust-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#adjust-counts" class="anchor"></a>Adjust <code>counts</code>
</h2>
<p>We may want to adjust <code>counts</code> for (known) unwanted variation. <code>adjust_abundance</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and a formula representing the desired linear model where the first covariate is the factor of interest and the second covariate is the unwanted variation, and returns a tibble with additional columns for the adjusted counts as <code>&lt;COUNT COLUMN&gt;_adjusted</code>. At the moment just an unwanted covariated is allowed at a time.</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="kw">tt.norm.adj</span> <span class="op">=</span>
    <span class="kw">tt.norm.batch</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="../reference/adjust_abundance.html">adjust_abundance</a></span>(
        <span class="op">~</span> <span class="kw">factor_of_interest</span> <span class="op">+</span> <span class="kw">batch</span>,
        .abundance = <span class="kw">count_scaled</span>,
        action = <span class="st">"only"</span>
      )

<span class="kw">tt.norm.adj</span>
</pre></div>
<pre><code>## # A tibble: 910 x 3
##    transcript sample     count_scaled_adjusted
##    &lt;chr&gt;      &lt;chr&gt;                      &lt;int&gt;
##  1 ACAP1      SRR1740034                 14644
##  2 ACP5       SRR1740034                  3903
##  3 ADRB2      SRR1740034                   719
##  4 AIM2       SRR1740034                  3707
##  5 ALOX5      SRR1740034                  7358
##  6 APOBEC3G   SRR1740034                  3104
##  7 APOL3      SRR1740034                  2417
##  8 APOL6      SRR1740034                  1110
##  9 ARHGAP22   SRR1740034                    69
## 10 BACH2      SRR1740034                 29826
## # … with 900 more rows</code></pre>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="kw">se.norm.batch</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/adjust_abundance.html">adjust_abundance</a></span>(
    <span class="op">~</span> <span class="kw">factor_of_interest</span> <span class="op">+</span> <span class="kw">batch</span>,
    .abundance = <span class="kw">count_scaled</span>
  )
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 19544 48 
## metadata(0):
## assays(3): count count_scaled count_scaled_adjusted
## rownames(19544): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
## rowData names(1): lowly_abundant
## colnames(48): SRR1740034 SRR1740035 ... SRR1740088 SRR1740089
## colData names(7): Cell.type time ... TMM multiplier</code></pre>
</div>
<div id="deconvolve-cell-type-composition" class="section level2">
<h2 class="hasAnchor">
<a href="#deconvolve-cell-type-composition" class="anchor"></a>Deconvolve <code>Cell type composition</code>
</h2>
<p>We may want to infer the cell type composition of our samples (with the algorithm Cibersort; Newman et al., 10.1038/nmeth.3337). <code>deconvolve_cellularity</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and returns a tibble with additional columns for the adjusted cell type proportions.</p>
<p><strong>columns truncated</strong></p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="kw">tt.cibersort</span> <span class="op">=</span>
    <span class="kw">tt</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/deconvolve_cellularity.html">deconvolve_cellularity</a></span>(action=<span class="st">"get"</span>, cores=<span class="fl">1</span>)

<span class="kw">tt.cibersort</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"cibersort:"</span>)) 
</pre></div>
<pre><code>## # A tibble: 5 x 23
##   sample `cibersort: B cells naive` `cibersort: B cells memory`
##   &lt;chr&gt;                       &lt;dbl&gt;                       &lt;dbl&gt;
## 1 SRR17…                    0.622                         0.238
## 2 SRR17…                    0.611                         0.257
## 3 SRR17…                    0                             0    
## 4 SRR17…                    0.00232                       0    
## 5 SRR17…                    0                             0    
## # … with 20 more variables: `cibersort: Plasma cells` &lt;dbl&gt;, `cibersort: T
## #   cells CD8` &lt;dbl&gt;, `cibersort: T cells CD4 naive` &lt;dbl&gt;, `cibersort: T cells
## #   CD4 memory resting` &lt;dbl&gt;, `cibersort: T cells CD4 memory activated` &lt;dbl&gt;,
## #   `cibersort: T cells follicular helper` &lt;dbl&gt;, `cibersort: T cells
## #   regulatory (Tregs)` &lt;dbl&gt;, `cibersort: T cells gamma delta` &lt;dbl&gt;,
## #   `cibersort: NK cells resting` &lt;dbl&gt;, `cibersort: NK cells activated` &lt;dbl&gt;,
## #   `cibersort: Monocytes` &lt;dbl&gt;, `cibersort: Macrophages M0` &lt;dbl&gt;,
## #   `cibersort: Macrophages M1` &lt;dbl&gt;, `cibersort: Macrophages M2` &lt;dbl&gt;,
## #   `cibersort: Dendritic cells resting` &lt;dbl&gt;, `cibersort: Dendritic cells
## #   activated` &lt;dbl&gt;, `cibersort: Mast cells resting` &lt;dbl&gt;, `cibersort: Mast
## #   cells activated` &lt;dbl&gt;, `cibersort: Eosinophils` &lt;dbl&gt;, `cibersort:
## #   Neutrophils` &lt;dbl&gt;</code></pre>
<p>With the new annotated data frame, we can plot the distributions of cell types across samples, and compare them with the nominal cell type labels to check for the purity of isolation. On the x axis we have the cell types inferred by Cibersort, on the y axis we have the inferred proportions. The data is facetted and coloured by nominal cell types (annotation given by the researcher after FACS sorting).</p>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="kw">tt.cibersort</span> <span class="op">%&gt;%</span>
    <span class="fu">gather</span>(<span class="kw">`Cell type inferred`</span>, <span class="kw">`proportion`</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">26</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/distinct.html">distinct</a></span>(<span class="kw">sample</span>, <span class="kw">`Cell type`</span>, <span class="kw">`Cell type inferred`</span>, <span class="kw">proportion</span>) <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">`Cell type inferred`</span>, y=<span class="kw">proportion</span>, fill=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span><span class="kw">`Cell type`</span>) <span class="op">+</span>
  <span class="kw">my_theme</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(axis.text.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(angle = <span class="fl">90</span>, hjust = <span class="fl">1</span>, vjust = <span class="fl">0.5</span>), aspect.ratio=<span class="fl">1</span><span class="op">/</span><span class="fl">5</span>)
</pre></div>
<p><img src="introduction_files/figure-html/plot_cibersort-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit">
<span class="kw">se.cibersort</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/deconvolve_cellularity.html">deconvolve_cellularity</a></span>(cores=<span class="fl">1</span>)
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 19544 48 
## metadata(0):
## assays(1): count
## rownames(19544): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
## rowData names(0):
## colnames(48): SRR1740034 SRR1740035 ... SRR1740088 SRR1740089
## colData names(27): Cell.type time ... cibersort..Eosinophils
##   cibersort..Neutrophils</code></pre>
</div>
<div id="cluster-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-samples" class="anchor"></a>Cluster <code>samples</code>
</h2>
<p>We may want to cluster our data (e.g., using k-means sample-wise). <code>cluster_elements</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and returns a tibble with additional columns for the cluster annotation. At the moment only k-means clustering is supported, the plan is to introduce more clustering methods.</p>
<p><strong>k-means</strong></p>
<div class="sourceCode" id="cb49"><pre class="downlit">
<span class="kw">tt.norm.cluster</span> <span class="op">=</span> <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(.abundance = <span class="kw">count_scaled</span>, method=<span class="st">"kmeans"</span>,  centers = <span class="fl">2</span> )

<span class="kw">tt.norm.cluster</span>
</pre></div>
<pre><code>## # A tibble: 2,635 x 12
##    sample transcript `Cell type` count time  condition `merged transcripts`
##    &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;                    &lt;dbl&gt;
##  1 SRR17… ABCB4      b_cell       1035 0 d   TRUE                         1
##  2 SRR17… ABCB9      b_cell         45 0 d   TRUE                         1
##  3 SRR17… ACAP1      b_cell       7151 0 d   TRUE                         1
##  4 SRR17… ACHE       b_cell          2 0 d   TRUE                         1
##  5 SRR17… ACP5       b_cell       2278 0 d   TRUE                         1
##  6 SRR17… ADAM28     b_cell      11156 0 d   TRUE                         1
##  7 SRR17… ADAMDEC1   b_cell         72 0 d   TRUE                         1
##  8 SRR17… ADAMTS3    b_cell          0 0 d   TRUE                         1
##  9 SRR17… ADRB2      b_cell        298 0 d   TRUE                         1
## 10 SRR17… AIF1       b_cell          8 0 d   TRUE                         1
## # … with 2,625 more rows, and 5 more variables: count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;,
## #   multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;, `cluster kmeans` &lt;fct&gt;</code></pre>
<p>We can add cluster annotation to the MDS dimesion reduced data set and plot.</p>
<div class="sourceCode" id="cb51"><pre class="downlit">
 <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(
    .abundance = <span class="kw">count_scaled</span>,
    method=<span class="st">"kmeans"</span>,
    centers = <span class="fl">2</span>,
    action=<span class="st">"get"</span>
  ) <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`cluster kmeans`</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/plot_cluster-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb52"><pre class="downlit">
<span class="kw">se.norm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(.abundance = <span class="kw">count_scaled</span>, method=<span class="st">"kmeans"</span>,  centers = <span class="fl">2</span> )
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 5 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(7): Cell.type time ... multiplier cluster.kmeans</code></pre>
<p><strong>SNN</strong></p>
<div class="sourceCode" id="cb54"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">=</span>   <span class="kw">tt.norm.tSNE</span> <span class="op">%&gt;%</span>    <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(.abundance= <span class="kw">count_scaled</span>, method = <span class="st">"SNN"</span>)
</pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 251
## Number of edges: 8484
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.5329
## Number of communities: 4
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/pivot_sample.html">pivot_sample</a></span>()
</pre></div>
<pre><code>## # A tibble: 251 x 5
##    sample                       Call    tSNE1  tSNE2 `cluster SNN`
##    &lt;chr&gt;                        &lt;fct&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;        
##  1 TCGA-A1-A0SD-01A-11R-A115-07 LumA    8.73   -4.55 0            
##  2 TCGA-A1-A0SF-01A-11R-A144-07 LumA   -6.01    4.90 2            
##  3 TCGA-A1-A0SG-01A-11R-A144-07 LumA    4.92   -9.92 1            
##  4 TCGA-A1-A0SH-01A-11R-A084-07 LumA    0.419  -5.60 0            
##  5 TCGA-A1-A0SI-01A-11R-A144-07 LumB    9.54   -1.31 0            
##  6 TCGA-A1-A0SJ-01A-11R-A084-07 LumA   -6.92   -1.58 1            
##  7 TCGA-A1-A0SK-01A-12R-A084-07 Basal -32.1    10.9  3            
##  8 TCGA-A1-A0SM-01A-11R-A084-07 LumA   -6.02  -12.0  2            
##  9 TCGA-A1-A0SN-01A-11R-A144-07 LumB   -4.75  -11.3  2            
## 10 TCGA-A1-A0SQ-01A-21R-A144-07 LumA   22.8    -5.18 1            
## # … with 241 more rows</code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit">
<span class="kw">tt.norm.SNN</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span>(<span class="st">"tSNE"</span>, ignore.case = <span class="fl">FALSE</span>), <span class="kw">`cluster SNN`</span>, <span class="kw">sample</span>, <span class="kw">Call</span>) <span class="op">%&gt;%</span>
    <span class="fu">gather</span>(<span class="kw">source</span>, <span class="kw">Call</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cluster SNN"</span>, <span class="st">"Call"</span>)) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/distinct.html">distinct</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">`tSNE1`</span>, y = <span class="kw">`tSNE2`</span>, color=<span class="kw">Call</span>)) <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(<span class="op">~</span><span class="kw">source</span>) <span class="op">+</span> <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/SNN-1.png" width="700"></p>
<div class="sourceCode" id="cb59"><pre class="downlit">
<span class="co"># Do differential transcription between clusters</span>
<span class="kw">tt.norm.SNN</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/mutate.html">mutate</a></span>(factor_of_interest = <span class="kw">`cluster SNN`</span> <span class="op">==</span> <span class="fl">3</span>) <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/test_differential_abundance.html">test_differential_abundance</a></span>(
    <span class="op">~</span> <span class="kw">factor_of_interest</span>,
    action=<span class="st">"only"</span>
   )
</pre></div>
<pre><code>## # A tibble: 500 x 8
##    ens           logFC logCPM     F   PValue      FDR significant lowly_abundant
##    &lt;chr&gt;         &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;lgl&gt;       &lt;lgl&gt;         
##  1 ENSG00000065…  1.59  10.2   395. 1.31e-53 6.56e-51 TRUE        FALSE         
##  2 ENSG00000111…  2.94   9.64  389. 4.51e-53 1.13e-50 TRUE        FALSE         
##  3 ENSG00000140…  2.69   9.54  345. 3.67e-49 6.12e-47 TRUE        FALSE         
##  4 ENSG00000186…  6.17   7.99  343. 5.74e-49 7.18e-47 TRUE        FALSE         
##  5 ENSG00000181…  7.93   9.12  299. 8.93e-45 8.93e-43 TRUE        FALSE         
##  6 ENSG00000137…  3.88   8.28  267. 1.85e-41 1.54e-39 TRUE        FALSE         
##  7 ENSG00000083…  1.37   9.39  247. 2.79e-39 1.99e-37 TRUE        FALSE         
##  8 ENSG00000143…  1.02  10.7   246. 3.56e-39 2.22e-37 TRUE        FALSE         
##  9 ENSG00000124…  4.66   8.62  238. 2.37e-38 1.32e-36 TRUE        FALSE         
## 10 ENSG00000092…  2.91   8.40  232. 1.36e-37 6.82e-36 TRUE        FALSE         
## # … with 490 more rows</code></pre>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="kw">se.norm.tSNE</span> <span class="op">%&gt;%</span>    <span class="fu"><a href="../reference/cluster_elements.html">cluster_elements</a></span>(.abundance= <span class="kw">count_scaled</span>, method = <span class="st">"SNN"</span>)
</pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 251
## Number of edges: 8484
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.5329
## Number of communities: 4
## Elapsed time: 0 seconds</code></pre>
<pre><code>## class: SummarizedExperiment 
## dim: 500 251 
## metadata(0):
## assays(2): count count_scaled
## rownames(500): ENSG00000002834 ENSG00000003989 ... ENSG00000265972
##   ENSG00000272398
## rowData names(0):
## colnames(251): TCGA-A1-A0SD-01A-11R-A115-07
##   TCGA-A1-A0SF-01A-11R-A144-07 ... TCGA-GM-A2DM-01A-11R-A180-07
##   TCGA-GM-A2DN-01A-11R-A180-07
## colData names(4): Call tSNE1 tSNE2 cluster.SNN</code></pre>
</div>
<div id="drop-redundant" class="section level2">
<h2 class="hasAnchor">
<a href="#drop-redundant" class="anchor"></a>Drop <code>redundant</code>
</h2>
<p>We may want to remove redundant elements from the original data set (e.g., samples or transcripts), for example if we want to define cell-type specific signatures with low sample redundancy. <code>remove_redundancy</code> takes as arguments a tibble, column names (as symbols; for <code>sample</code>, <code>transcript</code> and <code>count</code>) and returns a tibble dropped recundant elements (e.g., samples). Two redundancy estimation approaches are supported:</p>
<ul>
<li>removal of highly correlated clusters of elements (keeping a representative) with method=“correlation”</li>
<li>removal of most proximal element pairs in a reduced dimensional space.</li>
</ul>
<p><strong>Approach 1</strong></p>
<div class="sourceCode" id="cb64"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">=</span> <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="../reference/remove_redundancy.html">remove_redundancy</a></span>( method = <span class="st">"correlation"</span> )
</pre></div>
<p>We can visualise how the reduced redundancy with the reduced dimentions look like</p>
<div class="sourceCode" id="cb65"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/plot_drop-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb66"><pre class="downlit">
<span class="kw">se.norm.MDS</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="../reference/remove_redundancy.html">remove_redundancy</a></span>( method = <span class="st">"correlation"</span> )
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 4 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(4): SRR1740035 SRR1740043 SRR1740058 SRR1740067
## colData names(9): Cell.type time ... Dim2 Dim3</code></pre>
<p><strong>Approach 2</strong></p>
<div class="sourceCode" id="cb68"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">=</span>
    <span class="kw">tt.norm.MDS</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/remove_redundancy.html">remove_redundancy</a></span>(
    method = <span class="st">"reduced_dimensions"</span>,
    .element = <span class="kw">sample</span>,
    .feature = <span class="kw">transcript</span>,
    Dim_a_column = <span class="kw">`Dim1`</span>,
    Dim_b_column = <span class="kw">`Dim2`</span>
  )
</pre></div>
<p>We can visualise MDS reduced dimensions of the samples with the closest pair removed.</p>
<div class="sourceCode" id="cb69"><pre class="downlit">
<span class="kw">tt.norm.non_redundant</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/pivot_sample.html">pivot_sample</a></span>() <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x=<span class="kw">`Dim1`</span>, y=<span class="kw">`Dim2`</span>, color=<span class="kw">`Cell type`</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
  <span class="kw">my_theme</span>
</pre></div>
<p><img src="introduction_files/figure-html/plot_drop2-1.png" width="700"></p>
<p>All functions are also directly compatible with <code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit">
<span class="kw">se.norm.MDS</span> <span class="op">%&gt;%</span>
<span class="fu"><a href="../reference/remove_redundancy.html">remove_redundancy</a></span>(
    method = <span class="st">"reduced_dimensions"</span>,
    .element = <span class="kw">sample</span>,
    .feature = <span class="kw">transcript</span>,
    Dim_a_column = <span class="kw">`Dim1`</span>,
    Dim_b_column = <span class="kw">`Dim2`</span>
)
</pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 527 3 
## metadata(0):
## assays(2): count count_scaled
## rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
## rowData names(1): lowly_abundant
## colnames(3): SRR1740035 SRR1740058 SRR1740067
## colData names(9): Cell.type time ... Dim2 Dim3</code></pre>
</div>
<div id="other-useful-wrappers" class="section level2">
<h2 class="hasAnchor">
<a href="#other-useful-wrappers" class="anchor"></a>Other useful wrappers</h2>
<p>The above wrapper streamline the most common processing of bulk RNA sequencing data. Other useful wrappers are listed above.</p>
</div>
<div id="from-bamsam-to-tibble-of-gene-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#from-bamsam-to-tibble-of-gene-counts" class="anchor"></a>From BAM/SAM to tibble of gene counts</h2>
<p>We can calculate gene counts (using FeatureCounts; Liao Y et al., 10.1093/nar/gkz114) from a list of BAM/SAM files and format them into a tidy structure (similar to counts).</p>
<div class="sourceCode" id="cb72"><pre class="downlit">
<span class="kw">counts</span> <span class="op">=</span> <span class="fu"><a href="../reference/tidybulk_SAM_BAM.html">tidybulk_SAM_BAM</a></span>(
    <span class="kw">file_names</span>,
    genome = <span class="st">"hg38"</span>,
    isPairedEnd = <span class="fl">TRUE</span>,
    requireBothEndsMapped = <span class="fl">TRUE</span>,
    checkFragLength = <span class="fl">FALSE</span>,
    useMetaFeatures = <span class="fl">TRUE</span>
)
</pre></div>
</div>
<div id="from-ensembl-ids-to-gene-symbol-ids" class="section level2">
<h2 class="hasAnchor">
<a href="#from-ensembl-ids-to-gene-symbol-ids" class="anchor"></a>From ensembl IDs to gene symbol IDs</h2>
<p>We can add gene symbols from ensembl identifiers. This is useful since different resources use ensembl IDs while others use gene symbol IDs.</p>
<div class="sourceCode" id="cb73"><pre class="downlit">
<span class="kw">counts_ensembl</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/ensembl_to_symbol.html">ensembl_to_symbol</a></span>(<span class="kw">ens</span>)
</pre></div>
<pre><code>## # A tibble: 119 x 8
##    ens   iso   `read count` sample cases_0_project_disease_type
##    &lt;chr&gt; &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;                       
##  1 ENSG… 13             144 TARGE… Acute Myeloid Leukemia      
##  2 ENSG… 13              72 TARGE… Acute Myeloid Leukemia      
##  3 ENSG… 13               0 TARGE… Acute Myeloid Leukemia      
##  4 ENSG… 13            1099 TARGE… Acute Myeloid Leukemia      
##  5 ENSG… 13              11 TARGE… Acute Myeloid Leukemia      
##  6 ENSG… 13               2 TARGE… Acute Myeloid Leukemia      
##  7 ENSG… 13               3 TARGE… Acute Myeloid Leukemia      
##  8 ENSG… 13            2678 TARGE… Acute Myeloid Leukemia      
##  9 ENSG… 13             751 TARGE… Acute Myeloid Leukemia      
## 10 ENSG… 13               1 TARGE… Acute Myeloid Leukemia      
## # … with 109 more rows, and 3 more variables:
## #   cases_0_samples_0_sample_type &lt;chr&gt;, transcript &lt;chr&gt;, ref_genome &lt;chr&gt;</code></pre>
</div>
<div id="add-versus-get-versus-only-modes" class="section level2">
<h2 class="hasAnchor">
<a href="#add-versus-get-versus-only-modes" class="anchor"></a>ADD versus GET versus ONLY modes</h2>
<p>Every function takes a tidytranscriptomics structured data as input, and (i) with action=“add” outputs the new information joint to the original input data frame (default), (ii) with action=“get” the new information with the sample or transcript relative informatin depending on what the analysis is about, or (iii) with action=“only” just the new information. For example, from this data set</p>
<div class="sourceCode" id="cb75"><pre class="downlit">
  <span class="kw">tt.norm</span>
</pre></div>
<pre><code>## # A tibble: 2,635 x 11
##    sample transcript `Cell type` count time  condition `merged transcripts`
##    &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;                    &lt;dbl&gt;
##  1 SRR17… ABCB4      b_cell       1035 0 d   TRUE                         1
##  2 SRR17… ABCB9      b_cell         45 0 d   TRUE                         1
##  3 SRR17… ACAP1      b_cell       7151 0 d   TRUE                         1
##  4 SRR17… ACHE       b_cell          2 0 d   TRUE                         1
##  5 SRR17… ACP5       b_cell       2278 0 d   TRUE                         1
##  6 SRR17… ADAM28     b_cell      11156 0 d   TRUE                         1
##  7 SRR17… ADAMDEC1   b_cell         72 0 d   TRUE                         1
##  8 SRR17… ADAMTS3    b_cell          0 0 d   TRUE                         1
##  9 SRR17… ADRB2      b_cell        298 0 d   TRUE                         1
## 10 SRR17… AIF1       b_cell          8 0 d   TRUE                         1
## # … with 2,625 more rows, and 4 more variables: count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;,
## #   multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;</code></pre>
<p><strong>action=“add”</strong> (Default) We can add the MDS dimensions to the original data set</p>
<div class="sourceCode" id="cb77"><pre class="downlit">
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method=<span class="st">"MDS"</span> ,
        .element = <span class="kw">sample</span>,
        .feature = <span class="kw">transcript</span>,
        .dims = <span class="fl">3</span>,
        action=<span class="st">"add"</span>
    )
</pre></div>
<pre><code>## # A tibble: 2,635 x 14
##    sample transcript `Cell type` count time  condition `merged transcripts`
##    &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;                    &lt;dbl&gt;
##  1 SRR17… ABCB4      b_cell       1035 0 d   TRUE                         1
##  2 SRR17… ABCB9      b_cell         45 0 d   TRUE                         1
##  3 SRR17… ACAP1      b_cell       7151 0 d   TRUE                         1
##  4 SRR17… ACHE       b_cell          2 0 d   TRUE                         1
##  5 SRR17… ACP5       b_cell       2278 0 d   TRUE                         1
##  6 SRR17… ADAM28     b_cell      11156 0 d   TRUE                         1
##  7 SRR17… ADAMDEC1   b_cell         72 0 d   TRUE                         1
##  8 SRR17… ADAMTS3    b_cell          0 0 d   TRUE                         1
##  9 SRR17… ADRB2      b_cell        298 0 d   TRUE                         1
## 10 SRR17… AIF1       b_cell          8 0 d   TRUE                         1
## # … with 2,625 more rows, and 7 more variables: count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;,
## #   multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;, Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;, Dim3 &lt;dbl&gt;</code></pre>
<p><strong>action=“get”</strong> We can add the MDS dimensions to the original data set selecting just the sample-wise column</p>
<div class="sourceCode" id="cb79"><pre class="downlit">
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method=<span class="st">"MDS"</span> ,
        .element = <span class="kw">sample</span>,
        .feature = <span class="kw">transcript</span>,
        .dims = <span class="fl">3</span>,
        action=<span class="st">"get"</span>
    )
</pre></div>
<pre><code>## # A tibble: 5 x 10
##   sample `Cell type` time  condition `merged transcripts`   TMM multiplier
##   &lt;chr&gt;  &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;                    &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;
## 1 SRR17… b_cell      0 d   TRUE                         1 0.791       1.84
## 2 SRR17… b_cell      1 d   TRUE                         1 0.799       1.61
## 3 SRR17… monocyte    1 d   FALSE                        1 1.14        2.55
## 4 SRR17… t_cell      0 d   TRUE                         1 1.06        1.32
## 5 SRR17… dendritic_… 1 d   FALSE                        1 1.30        2.56
## # … with 3 more variables: Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;, Dim3 &lt;dbl&gt;</code></pre>
<p><strong>action=“only”</strong> We can get just the MDS dimensions relative to each sample</p>
<div class="sourceCode" id="cb81"><pre class="downlit">
  <span class="kw">tt.norm</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/reduce_dimensions.html">reduce_dimensions</a></span>(
        .abundance = <span class="kw">count_scaled</span>,
        method=<span class="st">"MDS"</span> ,
        .element = <span class="kw">sample</span>,
        .feature = <span class="kw">transcript</span>,
        .dims = <span class="fl">3</span>,
        action=<span class="st">"only"</span>
    )
</pre></div>
<pre><code>## # A tibble: 5 x 4
##   sample       Dim1    Dim2       Dim3
##   &lt;chr&gt;       &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
## 1 SRR1740034 -0.882 -0.708  -0.0291   
## 2 SRR1740035 -0.870 -0.715   0.0000434
## 3 SRR1740043  1.40   0.0479 -0.471    
## 4 SRR1740058 -0.935  1.39    0.00575  
## 5 SRR1740067  1.29  -0.0121  0.494</code></pre>
</div>
<div id="appendix" class="section level2">
<h2 class="hasAnchor">
<a href="#appendix" class="anchor"></a>Appendix</h2>
<div class="sourceCode" id="cb83"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
</pre></div>
<pre><code>## R version 4.0.2 Patched (2020-08-06 r78974)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-openmp/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] tidybulk_1.1.5   purrr_0.3.4      rlang_0.4.7      ggplot2_3.3.2   
##  [5] tidyr_1.1.1      magrittr_1.5     dplyr_1.0.1      tibble_3.0.3    
##  [9] knitr_1.29       BiocStyle_2.17.0
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.1.8             tidytext_0.2.5             
##   [3] plyr_1.8.6                  igraph_1.2.5               
##   [5] lazyeval_0.2.2              splines_4.0.2              
##   [7] BiocParallel_1.23.2         listenv_0.8.0              
##   [9] SnowballC_0.7.0             GenomeInfoDb_1.25.10       
##  [11] sva_3.37.0                  digest_0.6.25              
##  [13] htmltools_0.5.0             fansi_0.4.1                
##  [15] memoise_1.1.0               tensor_1.5                 
##  [17] cluster_2.1.0               ROCR_1.0-11                
##  [19] limma_3.45.9                globals_0.12.5             
##  [21] readr_1.3.1                 annotate_1.67.0            
##  [23] matrixStats_0.56.0          pkgdown_1.5.1.9000         
##  [25] colorspace_1.4-1            blob_1.2.1                 
##  [27] rappdirs_0.3.1              ggrepel_0.8.2              
##  [29] xfun_0.16                   crayon_1.3.4               
##  [31] RCurl_1.98-1.2              jsonlite_1.7.0             
##  [33] spatstat.data_1.4-3         spatstat_1.64-1            
##  [35] genefilter_1.71.0           survival_3.2-3             
##  [37] zoo_1.8-8                   ape_5.4                    
##  [39] glue_1.4.1                  polyclip_1.10-0            
##  [41] gtable_0.3.0                zlibbioc_1.35.0            
##  [43] XVector_0.29.3              leiden_0.3.3               
##  [45] DelayedArray_0.15.7         future.apply_1.6.0         
##  [47] BiocGenerics_0.35.4         abind_1.4-5                
##  [49] scales_1.1.1                DBI_1.1.0                  
##  [51] edgeR_3.31.4                miniUI_0.1.1.1             
##  [53] Rcpp_1.0.5                  widyr_0.1.3                
##  [55] viridisLite_0.3.0           xtable_1.8-4               
##  [57] reticulate_1.16             bit_4.0.4                  
##  [59] rsvd_1.0.3                  preprocessCore_1.51.0      
##  [61] stats4_4.0.2                htmlwidgets_1.5.1          
##  [63] httr_1.4.2                  RColorBrewer_1.1-2         
##  [65] ellipsis_0.3.1              Seurat_3.2.0               
##  [67] ica_1.0-2                   pkgconfig_2.0.3            
##  [69] XML_3.99-0.5                farver_2.0.3               
##  [71] uwot_0.1.8                  deldir_0.1-28              
##  [73] locfit_1.5-9.4              utf8_1.1.4                 
##  [75] tidyselect_1.1.0            labeling_0.3               
##  [77] reshape2_1.4.4              later_1.1.0.1              
##  [79] AnnotationDbi_1.51.3        munsell_0.5.0              
##  [81] tools_4.0.2                 cli_2.0.2                  
##  [83] generics_0.0.2              RSQLite_2.2.0              
##  [85] broom_0.7.0                 ggridges_0.5.2             
##  [87] evaluate_0.14               stringr_1.4.0              
##  [89] fastmap_1.0.1               goftest_1.2-2              
##  [91] yaml_2.2.1                  bit64_4.0.2                
##  [93] fs_1.5.0                    fitdistrplus_1.1-1         
##  [95] RANN_2.6.1                  pbapply_1.4-2              
##  [97] future_1.18.0               nlme_3.1-148               
##  [99] mime_0.9                    tokenizers_0.2.1           
## [101] compiler_4.0.2              plotly_4.9.2.1             
## [103] png_0.1-7                   e1071_1.7-3                
## [105] spatstat.utils_1.17-0       stringi_1.4.6              
## [107] desc_1.2.0                  lattice_0.20-41            
## [109] Matrix_1.2-18               vctrs_0.3.2                
## [111] pillar_1.4.6                lifecycle_0.2.0            
## [113] BiocManager_1.30.10         lmtest_0.9-37              
## [115] RcppAnnoy_0.0.16            data.table_1.13.0          
## [117] cowplot_1.0.0               bitops_1.0-6               
## [119] irlba_2.3.3                 httpuv_1.5.4               
## [121] patchwork_1.0.1             GenomicRanges_1.41.5       
## [123] R6_2.4.1                    bookdown_0.20              
## [125] promises_1.1.1              gridExtra_2.3              
## [127] KernSmooth_2.23-17          janeaustenr_0.1.5          
## [129] IRanges_2.23.10             codetools_0.2-16           
## [131] MASS_7.3-51.6               assertthat_0.2.1           
## [133] SummarizedExperiment_1.19.6 rprojroot_1.3-2            
## [135] withr_2.2.0                 sctransform_0.2.1          
## [137] S4Vectors_0.27.12           GenomeInfoDbData_1.2.3     
## [139] mgcv_1.8-31                 parallel_4.0.2             
## [141] hms_0.5.3                   rpart_4.1-15               
## [143] grid_4.0.2                  class_7.3-17               
## [145] rmarkdown_2.3               Rtsne_0.15                 
## [147] Biobase_2.49.0              shiny_1.5.0</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
